framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java

```diff
@@ -162,6 +162,29 @@
      */
     private boolean shouldDefaultTypeVarLocals;
 
+    /**
+     * Elements representing variables for which the type of the initializer is being determined in
+     * order to apply qualifier parameter defaults.
+     *
+     * <p>Local variables with a qualifier parameter get their declared type from the type of their
+     * initializer. Sometimes the initializer's type depends on the type of the variable, such as
+     * during type variable inference or when a variable is used in its own initializer as in
+     * "Object o = (o = null)". This creates a circular dependency resulting in infinite recursion.
+     * To prevent this, variables in this set should not be typed based on their initializer, but by
+     * using normal defaults.
+     *
+     * @see GenericAnnotatedTypeFactory#applyLocalVariableQualifierParameterDefaults
+     */
+    private Set<Element> variablesUnderInitialization;
```

Comment:
Can you use the more precise `VariableElement`?

Reply:
The documentation on this seems sketchy, the [javadoc](https://docs.oracle.com/javase/8/docs/api/javax/lang/model/element/ElementKind.html#LOCAL_VARIABLE) for this doesn't seem guarantee that `e.getKind() == ElementKind.LOCAL_VARIABLE` implies `e instanceof VariableElement`. However, the only line in the jdk source code `return ElementKind.LOCAL_VARIABLE` is within a `VariableElement` implementation, so I think your suggestion makes sense.